In [ ]:
import pandas as pd
import os
import pickle

data_loc = "C:/Users/jessica52/OneDrive - Discovery/WORKSTUFF/DiscoverIt/Qhack/Hack data sharing/"
model_loc = "C:/Users/jessica52/OneDrive - Discovery/WORKSTUFF/DiscoverIt/Qhack/llm-hackathon/models/"
python_backup_loc = "C:/Users/jessica52/OneDrive - Discovery/WORKSTUFF/DiscoverIt/Qhack/llm-hackathon/python_backup/"

#OPENAI_API_KEY = "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"
#os.environ["OPENAI_API_KEY"] = OPENAI_API_KEY

from langchain.document_loaders import PyPDFDirectoryLoader
from langchain.document_loaders import UnstructuredMarkdownLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings import HuggingFaceEmbeddings, SentenceTransformerEmbeddings
from langchain.llms import GPT4All
from langchain.chains import RetrievalQA
from langchain.vectorstores import Qdrant
from langchain.prompts import PromptTemplate
from langchain.agents import Tool
from langchain.agents import initialize_agent
from langchain.agents import AgentType
#from langchain.vectorstores import Chroma # Seems to need MS developer tools

In [ ]:
# https://python.langchain.com/docs/use_cases/question_answering/ !!!!

# Read pdf and md files

OPTIONS for default data loading
* different data loaders/connectors (120 options)
* the PyPDFDirectoryLoader used automatically splits pages - we may not want this. TODO: investigate

TODO:
* read tables
* hierarchy of headers etc.

In [23]:
content_data_loc = data_loc + "Knowledge content data"
reload_content_data = False

if reload_content_data:
    md_files = [f for f in os.listdir(content_data_loc) if f.endswith('.md')]
    md_docs = []
    for f in md_files:
        md_docs.extend(UnstructuredMarkdownLoader(content_data_loc +"/"+f).load())

    pdf_loader = PyPDFDirectoryLoader(content_data_loc) # automatically splits pages
    pdf_docs = pdf_loader.load() # takes longish

    all_docs = pdf_docs + md_docs

    with open(python_backup_loc + "all_docs", 'wb') as f:
        pickle.dump(all_docs, f)

else:
    with open(python_backup_loc + "all_docs", 'rb') as f:
        all_docs = pickle.load(f)

print(len(all_docs))

1166


In [30]:
all_docs[0]

Document(page_content=' \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \nDiscovery Health Medical Scheme, registration number 1125, is regulated by the Council for Medical Schemes and administered  by Discovery Health (Pty) Ltd, registration \nnumber 1997/013480/07, an authorised financial services provider.  \n ADVANCED ILLNESS  \nBENEFIT  \n \nDISCOVERY HEALTH MEDICAL SCHEME  \n2022  \n ', metadata={'source': 'C:\\Users\\jessica52\\OneDrive - Discovery\\WORKSTUFF\\DiscoverIt\\Qhack\\Hack data sharing\\Knowledge content data\\Benefit Guides\\Benefits\\advanced_illness_benefit 2022.pdf', 'page': 0})

In [33]:
for d in all_docs:
    if 'chronic_illness_benefit_medicine_list_formulary 2022' in d.metadata['source']:
        if  d.metadata['page'] == 3:
            print(d.page_content)

 
 
 
 
 
 
 
Discovery Health Medical Scheme, registration number 1125, is regulated by the Council for Medical Schemes and administered  by Discovery Health (Pty) Ltd,  
registration number 1997/013480/07, an authorised financial services provider . Find a healthcare provider and the Discovery app are brought to you by Discovery Health (Pty) 
Ltd; registration number 1997/013480/07, an authorised financial services provider and administrator of medical schemes.    4 
 
 
 
 
 The Prescribed Minimum Benefits (PMBs) Chronic Disease List (CDL) medicine list for all plan 
types  
The PMB CDL medicine list applies to all Discovery Health Medical Scheme plans for the 2022 benefit year. The Chronic Drug Amount (CDA) 
does not apply to the KeyCare and Smart plans.  
 
Medicine marked with an asterisk (*) will soon be removed from the medic ine list (formulary)  
 
 
CHRONIC 
DISEASE LIST 
CONDITION  CDA FOR  
CORE, 
PRIORITY  
AND  
SAVER  
PLANS  CDA FOR  
EXECUTIVE  AND  
COMPREHENSIVE  
P

# Clean document content TODO

* new lines
* Discovery Health Medical Scheme, registration number 1125, is regulated by the Council for Medical Schemes and administered  by Discovery Health (Pty) Ltd, registration \nnumber 1997/013480/07, an authorised financial services provider
* page numbers

# Split content

OPTIONS for splitting:
* https://python.langchain.com/docs/modules/data_connection/document_transformers/
* https://python.langchain.com/docs/use_cases/question_answering/document-context-aware-QA
* would like to consider context-aware splitting
    * https://python.langchain.com/docs/modules/data_connection/document_loaders/integrations/grobid (seems like a pain to install)


In [25]:
text_splitter = RecursiveCharacterTextSplitter(chunk_size = 500, chunk_overlap = 100)
all_doc_splits = text_splitter.split_documents(all_docs)# + pdf_docs 

print(len(all_doc_splits))

7053


# Load into vector database

TODO: token length

OPTIONS for embedding model:
* huggingface models e.g. "sentence-transformers/all-mpnet-base-v2"
* OpenAIEmbeddings()

OPTIONS for vector store:
* gone with Qdrant for now

In [26]:
hfe = HuggingFaceEmbeddings(
    model_name="sentence-transformers/all-mpnet-base-v2",
    model_kwargs={'device': 'cpu'},
    encode_kwargs={'normalize_embeddings': False}
)

In [27]:
qdrant = Qdrant.from_documents(
    all_doc_splits,
    hfe,
    location=":memory:",  # Local mode with in-memory storage only
    # for on-disk, use path = "path"
    collection_name="my_documents",
)

# Answer query - contextless

In [ ]:
llm = GPT4All(model=model_loc+"nous-hermes-13b.ggmlv3.q4_0.bin",max_tokens=2048)

In [ ]:
llm = GPT4All(model=model_loc+"ggml-model-gpt4all-falcon-q4_0.bin",max_tokens=2048)

In [35]:
query = "what is the above threshold benefit?"

## Similarity matching

OPTIONS: 
* similarity search - default distance metric?
* max_marginal_relevance_search
* SVM retriever https://twitter.com/karpathy/status/1647025230546886658?s=20
* Multi-query retriever https://python.langchain.com/docs/modules/data_connection/retrievers/how_to/MultiQueryRetriever

In [26]:
found_docs = qdrant.similarity_search_with_score(query)
document, score = found_docs[0]
print(document.page_content)
print(f"\nScore: {score}")

Above Threshold Benefit

Available on the Executive, Comprehensive and Priority plans

Score: 0.7556167658706837


## Meta-data filtering TODO

Qdrant: https://qdrant.tech/documentation/concepts/filtering/

https://python.langchain.com/docs/modules/data_connection/vectorstores/integrations/qdrant#metadata !

In [ ]:
#from qdrant_client.http import models as rest
#found_docs = qdrant.similarity_search_with_score(query, filter=rest.Filter(...))

## Using generative LLMs basics

OPTIONS
* https://blog.langchain.dev/chat-models/
* https://python.langchain.com/docs/modules/model_io/models/chat/
* llm = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0)
* GPT4All
    * models have to be downloaded from https://gpt4all.io/index.html See "Model Explorer"
    * nous-hermes-13b.ggmlv3.q4_0.bin # best overall can't be used commerically
    * ggml-model-gpt4all-falcon-q4_0.bin # best smaller, commercial use.

In [36]:
qa_chain = RetrievalQA.from_chain_type(llm,retriever=qdrant.as_retriever(), return_source_documents=True)
result = qa_chain({"query": query})

print(result.result)

Found model file at  C:/Users/jessica52/OneDrive - Discovery/WORKSTUFF/DiscoverIt/Qhack/llm-hackathon/models/nous-hermes-13b.ggmlv3.q4_0.bin


AttributeError: 'dict' object has no attribute 'result'

## Customising how retrievers are used with generative LLMs TODO

stuff, refine, map-reduce, and map-rerank chains for passing documents to an LLM prompt are well summarized https://python.langchain.com/docs/modules/chains/document/.


In [ ]:
from langchain.chains.question_answering import load_qa_chain
chain = load_qa_chain(llm, chain_type="stuff")
chain({"input_documents": found_docs, "query": query},return_only_outputs=True)

qa_chain = RetrievalQA.from_chain_type(llm,retriever=qdrant.as_retriever(),
                                       chain_type="stuff")
result = qa_chain({"query": query})

## Using generative LLMs with prompt engineering

In [36]:
template = """Use the following pieces of context to answer the question at the end. 
If you don't know the answer, just say that you don't know, don't try to make up an answer. 
Use three sentences maximum and keep the answer as concise as possible. 
{context}
Question: {question}
Helpful Answer:"""
QA_CHAIN_PROMPT = PromptTemplate(input_variables=["context", "query"],template=template,)

# Run chain

qa_chain = RetrievalQA.from_chain_type(llm,
                                       retriever=qdrant.as_retriever(),
                                       chain_type_kwargs={"prompt": QA_CHAIN_PROMPT},
                                      return_source_documents=True)

result = qa_chain({"query": query})
print(result["result"])
print(len(result['source_documents'])) # Returning source documents
print(result['source_documents'][0])


Found model file at  C:/Users/jessica52/OneDrive - Discovery/WORKSTUFF/DiscoverIt/Qhack/llm-hackathon/models/ggml-model-gpt4all-falcon-q4_0.bin


' The Above Thr'

## Using generative LLMs with citations

In [ ]:
# Citations
from langchain.chains import RetrievalQAWithSourcesChain
qa_chain = RetrievalQAWithSourcesChain.from_chain_type(llm,retriever=qdrant.as_retriever())
result = qa_chain({"query": query})
result

# Guard query and result

# Bringing in client context!

In [46]:
this_entity = '26-anon'

In [41]:
xls = pd.ExcelFile(data_loc + 'Client context data/client_context_anon.xlsx')
state = pd.read_excel(xls, 'client state')
claims = pd.read_excel(xls, 'client claims')

In [ ]:
def get_field(field_name: str, state_data: pd.DataFrame, entity_no:str):
    """Returns the specified field from the state_data DataFrame for a given entity unmber."""
    try:
        result = state_data[state_data['ENTITY_NO'] == entity_no][field_name][0]
        if result == "[NULL]":
            return None
        else: 
            return result
    except KeyError:
        return None
    
def get_plan(state_data: pd.DataFrame, entity_no:str) -> str:
    "Returns the medical aid plan description for the provided entity"
    return get_field("PLAN", state_data, entity_no)


def get_plan_dummy():
    return "Classic Smart"


tools = [
    Tool(
        name="get plan",
        func=get_plan_dummy,
        #description="useful for when you need to get the medical aid plan description for a provided entity and set of state data"
        description="useful for when you need to get the medical aid plan description
    )
]

In [38]:
query1 = "how many GP visits are covered for my baby on the Smart plan?"
query2 = "how many GP visits are covered for my baby on my plan?"

In [39]:
qa_chain = RetrievalQA.from_chain_type(llm,retriever=qdrant.as_retriever(), return_source_documents=True)
result = qa_chain({"query": query1})

print(result['result'])

Found model file at  C:/Users/jessica52/OneDrive - Discovery/WORKSTUFF/DiscoverIt/Qhack/llm-hackathon/models/ggml-model-gpt4all-falcon-q4_0.bin


AttributeError: 'dict' object has no attribute 'result'

In [62]:
print(result['result'])

 Your baby under the age of two years is covered for two visits to a GP, paeediactrician or an ear, nose and throat specialist.


In [64]:
query_with_context = query +" "+ get_plan(state, this_entity)
query_with_context

'what is the above threshold benefit? Classic Smart'

In [61]:
result

{'query': 'how many GP visits are covered for my baby on the Smart plan?',
 'result': ' Your baby under the age of two years is covered for two visits to a GP, paeediactrician or an ear, nose and throat specialist.',
 'source_documents': [Document(page_content='do not have day -to-day benefits, or if you have run out of funds, you will have to pay for these costs. Please refer to the Maternity \nbenefits available for your plan type section for more details.  \nAntenatal consultations  \nOnce activated, you are covered for between 8 to 12 visits (including the urine dipstick) at your gynaecologist, GP or midwif e, \ndepending on the plan you choose from the date of activation. If you are on a Classic Smart Comprehensive, Classic Smart,', metadata={'source': 'C:\\Users\\jessica52\\OneDrive - Discovery\\WORKSTUFF\\DiscoverIt\\Qhack\\Hack data sharing\\Knowledge content data\\Benefit Guides\\Benefits\\cover_for_pregnancy_childbirth_and_early_childhood 2022.pdf', 'page': 3}),
  Document(pa

In [ ]:
agent = initialize_agent(tools, llm, agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose=True, retriever=qdrant.as_retriever())
agent.run(query2)

Found model file at  C:/Users/jessica52/OneDrive - Discovery/WORKSTUFF/DiscoverIt/Qhack/llm-hackathon/models/ggml-model-gpt4all-falcon-q4_0.bin


> Entering new  chain...
 I should check the medical aid plan description for my baby's entity and state data
Action: use the `get_plan` function with the entity and state data
Action Input: `entity_id` and `state_id`
Observation: use the `get_plan` function with the entity and state data is not a valid tool, try another one.
Thought: I should check the medical aid plan description for my baby's entity and state data
Action: use the `get_plan` function with the entity and state data
Action Input: `entity_id` and `state_id`
Observ
Observation: use the `get_plan` function with the entity and state data is not a valid tool, try another one.
Thought: I should check the medical aid plan description for my baby's entity and state data
Action: use the `get_plan` function with the entity and state data
Action Input: `entity_id` and `state_id`
Observ


Exception ignored on calling ctypes callback function: <function LLModel._prompt_callback at 0x000001E03866F1C0>
Traceback (most recent call last):
  File "C:\Users\jessica52\Anaconda3\envs\llm_hack\lib\site-packages\gpt4all\pyllmodel.py", line 368, in _prompt_callback
    @staticmethod
KeyboardInterrupt: 
Exception ignored on calling ctypes callback function: <function LLModel._prompt_callback at 0x000001E03866F1C0>
Traceback (most recent call last):
  File "C:\Users\jessica52\Anaconda3\envs\llm_hack\lib\site-packages\gpt4all\pyllmodel.py", line 368, in _prompt_callback
    @staticmethod
KeyboardInterrupt: 


 I should check the medical aid plan description for my baby's entity and state data
Action: use the `get_plan` function with the entity and state data
Action Input: `entity_id` and `state_id`
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Obse
Observation: use the `get_plan` function with the entity and state data is not a valid tool, try another one.
Thought: I should check the medical aid plan description for my baby's entity and state data
Action: use the `get_plan` function with the entity and state data
Action Input: `entity_id` and `state_id`
Observation: use the `get_plan` function with the entity and state data is not a valid tool, try another one.
Thought:

Exception ignored on calling ctypes callback function: <function LLModel._response_callback at 0x000001E03866F250>
Traceback (most recent call last):
  File "C:\Users\jessica52\Anaconda3\envs\llm_hack\lib\site-packages\gpt4all\pyllmodel.py", line 373, in _response_callback
    @staticmethod
KeyboardInterrupt: 
Exception ignored on calling ctypes callback function: <function LLModel._response_callback at 0x000001E03866F250>
Traceback (most recent call last):
  File "C:\Users\jessica52\Anaconda3\envs\llm_hack\lib\site-packages\gpt4all\pyllmodel.py", line 373, in _response_callback
    @staticmethod
KeyboardInterrupt: 
Exception ignored on calling ctypes callback function: <function LLModel._response_callback at 0x000001E03866F250>
Traceback (most recent call last):
  File "C:\Users\jessica52\Anaconda3\envs\llm_hack\lib\site-packages\gpt4all\pyllmodel.py", line 373, in _response_callback
    @staticmethod
KeyboardInterrupt: 
Exception ignored on calling ctypes callback function: <function